### PR TITLE
Add MaxDepth to limit recursion

### PR DESCRIPTION
**Pull request by thockin**
@lavalamp I'm using fuzz for something and needed this

### CODE REVIEW

File: fuzz.go [hunk 1]
```diff
@@ -179,6 +192,12 @@ const (
 )
 
 func (f *Fuzzer) doFuzz(v reflect.Value, flags uint64) {
+	if f.curDepth >= f.maxDepth {
+		return
+	}
+	f.curDepth++
```
Comment (lavalamp) [inline fuzz.go]: This makes the fuzzer non-thread-safe...

Comment (lavalamp) [inline fuzz.go]: I would probably make another struct containing curDepth and the *Fuzzer to use, construct this object in the Fuzz method, and implement doFuzz on this new object.

Comment (thockin) [inline fuzz.go]: Do we care?  The alternative is to carry around a context object up and down the call stack?

**Comment by thockin** [comment]
PTAL - passes `go test` and makes a new fuzzerContext to carry depth.

File: fuzz.go [hunk 2]
```diff
@@ -148,17 +158,19 @@ func (f *Fuzzer) genShouldFill() bool {
 // fails, this will generate random values for all primitive fields and then
 // recurse for all non-primitives.
 //
-// Not safe for cyclic or tree-like structs!
+// This is safe for cyclic or tree-like structs, up to a limit.  Use use the
```
Comment (lavalamp) [inline fuzz.go]: nit: s/Use use/Use/

**Comment by thockin** [comment]
nit fixed

**Comment by lavalamp** [comment]
LGTM

**Comment by thockin** [comment]
merge?  Or add me as a repo owner? :)

**Comment by lavalamp** [comment]
Was waiting for test :)

**Pull request merged**